File: src/harness/adapters/continue_cli.py
"""continue-cli adapter — invokes the `cn` CLI (Continue) in print mode.

When OPENAI-style env vars are present, this adapter writes a minimal Continue
config YAML and runs `cn -p --config <file> --format json ...` so bare model
IDs like `gpt-5.4` work against an OpenAI-compatible endpoint.
"""
from __future__ import annotations

import json
import os
import re
from pathlib import Path

from harness._subproc import SubprocOutcome, write_instructions
from harness.base import (
    Adapter,
    AgentStatus,
    BuildCommand,
    InstallMeta,
    ParsedOutput,
    ReadyState,
    RunSpec,
    SessionTelemetry,
)
from harness.pricing import derive_cost
from harness.util import last_non_empty_join, strip_ansi

_MODEL_LOADING_RE = re.compile(r"Model:\s*Loading", re.IGNORECASE)
_ASK_ANYTHING_RE = re.compile(r"Ask anything", re.IGNORECASE)
_UPDATE_RE = re.compile(r"Update available", re.IGNORECASE)
_RATE_LIMIT_RE = re.compile(r"rate.?limit", re.IGNORECASE)
_SPINNER_RE = re.compile(r"[⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏]")
_WORKING_RE = re.compile(r"thinking|working", re.IGNORECASE)


class ContinueCliAdapter(Adapter):
    name = "continue-cli"
    instructions_filename = "CONTINUE.md"
    submit_keys = ("Enter",)
    install_meta = InstallMeta(
        package_manager="npm",
        install_command=("npm", "install", "-g", "@continuedev/cli"),
        update_command=("npm", "install", "-g", "@continuedev/cli@latest"),
        version_command=("cn", "--version"),
    )

    DEFAULT_MODEL = "claude-sonnet-4-6"

    def build_command(self, spec: RunSpec) -> BuildCommand:
        resolved = self.resolve_run_spec(spec)
        model = resolved.model
        instructions_file = write_instructions(spec.workdir, self.instructions_filename, spec.instructions)

        openai_key = spec.env.get("OPENAI_API_KEY")
        openai_base = spec.env.get("OPENAI_BASE_URL")
        if openai_key or openai_base:
            config_path = _write_continue_config(Path(spec.workdir), model, openai_key or "dummy", openai_base, spec.instructions)
            args = ["-p", "--config", str(config_path), "--format", "json", spec.prompt]
            return BuildCommand(cmd="cn", args=args, cwd=spec.workdir, env={}, instructions_file=instructions_file)

        args = ["-p", spec.prompt, "--model", model, "--json"]
        return BuildCommand(cmd="cn", args=args, cwd=spec.workdir, env={}, instructions_file=instructions_file)

    def parse_output(self, spec: RunSpec, outcome: SubprocOutcome) -> ParsedOutput:
        raw: dict | None = None
        if outcome.stdout.strip():
            try:
                raw = json.loads(outcome.stdout)
            except json.JSONDecodeError:
                raw = None

        cost = tokens_in = tokens_out = None
        if isinstance(raw, dict):
            usage = raw.get("usage") or {}
            tokens_in = usage.get("input_tokens")
            tokens_out = usage.get("output_tokens")
            cost = raw.get("total_cost_usd")

        return {"cost_usd": cost, "tokens_in": tokens_in, "tokens_out": tokens_out, "raw": raw}

    def session_log_path(self, workdir: Path, session_started_after: float | None = None) -> str | None:
        env_dir = os.environ.get("CONTINUE_SESSION_DIR")
        if env_dir:
            d = Path(env_dir).expanduser()
            if d.exists() and d.is_dir():
                files = sorted((p for p in d.glob("*.json") if p.is_file()), key=lambda p: p.stat().st_mtime, reverse=True)
                if files:
                    return str(files[0])

        base = workdir.name
        candidates = [
            Path.home() / ".continue" / "sessions" / base,
            Path.home() / ".continue" / "dev_data" / base,
            Path.home() / ".continue" / "index" / base,
        ]
        for d in candidates:
            if not d.exists() or not d.is_dir():
                continue
            files = sorted((p for p in d.glob("*.json") if p.is_file()), key=lambda p: p.stat().st_mtime, reverse=True)
            if files:
                return str(files[0])
            idx = d / "session.json"
            if idx.exists():
                return str(idx)
        return None

    def parse_session_log(self, path: str) -> SessionTelemetry:
        p = Path(path)
        if not p.exists():
            return SessionTelemetry(path, None, None, None, None, None)
        try:
            raw = json.loads(p.read_text(encoding="utf-8"))
            obj = raw if isinstance(raw, dict) else {}
            usage = obj.get("usage") if isinstance(obj.get("usage"), dict) else {}
            tokens_in = int(usage.get("input_tokens")) if isinstance(usage.get("input_tokens"), (int, float)) else None
            tokens_out = int(usage.get("output_tokens")) if isinstance(usage.get("output_tokens"), (int, float)) else None
            model = obj.get("model") if isinstance(obj.get("model"), str) else None
            cost = float(obj.get("total_cost_usd")) if isinstance(obj.get("total_cost_usd"), (int, float)) else None
            if cost is None:
                cost = derive_cost(model, tokens_in, tokens_out)
            return SessionTelemetry(path, tokens_in, tokens_out, cost, model, raw)
        except (OSError, json.JSONDecodeError, ValueError, TypeError):
            return SessionTelemetry(path, None, None, None, None, None)

    # ---- session-aware ---------------------------------------------------

    def detect_ready(self, pane: str) -> ReadyState:
        last20 = last_non_empty_join(pane, 20)
        # cn shows "Ask anything" placeholder while the model is still loading.
        # Real ready = input visible AND model loaded (no "Model: Loading..." line).
        if _MODEL_LOADING_RE.search(last20):
            return "loading"
        if _ASK_ANYTHING_RE.search(last20):
            return "ready"
        if _UPDATE_RE.search(last20):
            return "dialog"
        return "loading"

    def handle_dialog(self, pane: str) -> list[str] | None:
        if _UPDATE_RE.search(strip_ansi(pane)):
            return ["Escape"]
        return None

    def detect_status(self, pane: str) -> AgentStatus:
        last10 = last_non_empty_join(pane, 10)
        if _RATE_LIMIT_RE.search(last10):
            return "rate-limited"
        if _SPINNER_RE.search(last10) or _WORKING_RE.search(last10):
            return "running"
        if _ASK_ANYTHING_RE.search(last10):
            return "idle"
        return "unknown"


def _write_continue_config(workdir: Path, model: str, api_key: str, api_base: str | None, instructions: str | None) -> Path:
    continue_dir = workdir / ".harness" / "continue"
    continue_dir.mkdir(parents=True, exist_ok=True)
    config_path = continue_dir / "config.yaml"
    lines = [
        "name: Harness Continue",
        "version: 1.0.0",
        "schema: v1",
        "models:",
        "  - name: harness-model",
        f"    model: {model}",
        "    provider: openai",
        f"    apiKey: {api_key}",
    ]
    if api_base:
        lines.append(f"    apiBase: {api_base}")
    lines += [
        "    roles:",
        "      - chat",
        "      - edit",
        "      - apply",
    ]
    if instructions:
        escaped = instructions.rstrip().replace("\n", "\\n")
        lines += ["rules:", f"  - '{escaped}'"]
    config_path.write_text("\n".join(lines) + "\n", encoding="utf-8")
    return config_path

File: ts/src/adapters/continue-cli.ts
import { register } from '../registry.js'
import type { Adapter, AgentStatus, BuildCommand, ParsedOutput, ReadyState, RunSpec, SessionTelemetry, SubprocOutcome } from '../base.js'
import { HarnessError, finalizeCommand, validateRunSpec } from '../base.js'
import { stripAnsi, lastNonEmptyJoin } from '../util.js'
import { readFileSync, readdirSync, realpathSync, statSync } from 'node:fs'
import type { Dirent } from 'node:fs'
import { join, resolve } from 'node:path'
import { homedir } from 'node:os'

/**
 * Session artifacts (continuedev/continue@5522c6f, `extensions/cli/src/session.ts`
 * and `core/util/{history,paths}.ts`): `<CONTINUE_GLOBAL_DIR or ~/.continue>/sessions/
 * <uuid>.json` holding `{sessionId, title, workspaceDirectory, history, usage?}`
 * where `workspaceDirectory` is the `cn` process cwd and `usage` is the CLI's own
 * cumulative `{totalCost, promptTokens, completionTokens, ...}`. `sessions.json`
 * in the same directory is the upstream index, not a session. The CLI never
 * persists a model id (`chatModelTitle` is unset), so `model` stays null.
 */

/** `<CONTINUE_GLOBAL_DIR or ~/.continue>/sessions`; a relative override resolves against the process cwd. */
export function continueSessionsDir(): string {
  const override = process.env['CONTINUE_GLOBAL_DIR']
  return join(override ? resolve(override) : join(process.env['HOME'] || homedir(), '.continue'), 'sessions')
}

function asRecord(value: unknown): Record<string, unknown> | null {
  return value !== null && typeof value === 'object' && !Array.isArray(value) ? value as Record<string, unknown> : null
}

function readJsonObject(path: string): Record<string, unknown> | null {
  try { return asRecord(JSON.parse(readFileSync(path, 'utf8'))) } catch { return null }
}

function continueSessionPath(workdir: string, since?: number): string | null {
  const absolute = resolve(workdir)
  let resolved = absolute
  try { resolved = realpathSync(absolute) } catch { /* keep the unresolved path */ }
  // Distinct case-sensitive POSIX paths must not share telemetry.

  const dir = continueSessionsDir()
  let entries: Dirent[]
  try { entries = readdirSync(dir, { withFileTypes: true }) } catch { return null }
  const candidates: { mtimeMs: number; path: string }[] = []
  for (const entry of entries) {
    if (!entry.isFile() || !entry.name.endsWith('.json') || entry.name === 'sessions.json') continue
    const path = join(dir, entry.name)
    let mtimeMs: number
    try { mtimeMs = statSync(path).mtimeMs } catch { continue }
    if (since !== undefined && mtimeMs < since) continue
    candidates.push({ mtimeMs, path })
  }
  candidates.sort((a, b) => b.mtimeMs - a.mtimeMs || (a.path < b.path ? 1 : a.path > b.path ? -1 : 0))
  for (const { path } of candidates) {
    const workspace = readJsonObject(path)?.['workspaceDirectory']
    if (workspace === absolute || workspace === resolved) return path
  }
  return null
}

function numberOrNull(v: unknown): number | null {
  return typeof v === 'number' && Number.isFinite(v) ? v : null
}

/** `cn --model` takes a Continue Hub slug: exactly `owner/package`. */
function isHubSlug(model: string): boolean {
  const slash = model.indexOf('/')
  return slash > 0 && slash < model.length - 1 && !model.includes('/', slash + 1)
}

/**
 * A standalone `{status, message, ...}` line `cn` prints to stdout while
 * (auto-)compacting before the final response (commands/chat.ts).
 */
function isCompactionStatusLine(line: string): boolean {
  if (!line.startsWith('{')) return false
  try {
    const obj: unknown = JSON.parse(line)
    return typeof obj === 'object' && obj !== null && !Array.isArray(obj) && typeof (obj as Record<string, unknown>)['status'] === 'string'
  } catch {
    return false
  }
}

/**
 * The final JSON `cn --format json` prints: the model text verbatim when it
 * parsed as JSON upstream, else the `{response, status, note}` wrapper.
 * Leading compaction status lines are skipped; anything else is null.
 */
function parseHeadlessJson(stdout: string): unknown | null {
  const text = stdout.trim()
  if (!text) return null
  try {
    return JSON.parse(text)
  } catch {
    // fall through to compaction-prefix handling
  }
  const lines = text.split('\n')
  let start = 0
  while (start < lines.length - 1 && isCompactionStatusLine(lines[start]!.trim())) start++
  if (start === 0) return null
  try {
    return JSON.parse(lines.slice(start).join('\n'))
  } catch {
    return null
  }
}

/**
 * Continue headless adapter; see SPEC.md and ADAPTER-MATRIX.md.
 * --model selects a Hub slug; an empty default delegates to upstream config.
 * Projected instructions require --rule. JSON is model output, not telemetry.
 */
const continueCliAdapter: Adapter = {
  name: 'continue-cli',
  instructionsFilename: 'CONTINUE.md',
  /** Empty means "upstream selection": no `--model`, reported model null. */
  defaultModel: '',
  permissionBypassArgs: ['--auto'],
  configFileFlag: '--config',

  buildCommand(spec: RunSpec): BuildCommand {
    const validated = validateRunSpec(this, spec)
    const { model, configArgs, configFile, permissionArgs } = validated
    // Same path finalizeCommand plans for the projection; passed as a rule
    // because cn never reads a root CONTINUE.md.
    const ruleArgs = spec.instructions === undefined ? [] : ['--rule', join(validated.workdir, this.instructionsFilename)]

    if (configFile !== null) {
      // Any explicit model (even whitespace) is rejected; empty/omitted defers to the file like Python.
      if (spec.model) {
        throw new HarnessError(
          'continue-cli configFile selects the model itself (cn --model is a Hub slug); omit spec.model or drop configFile',
          'unsupported-capability',
        )
      }
      return finalizeCommand(this, spec, validated, {
        cmd: 'cn',
        args: ['-p', ...configArgs, ...permissionArgs, ...ruleArgs, '--format', 'json', spec.prompt],
        model: null,
      })
    }

    const env = spec.env ?? {}
    if (env['OPENAI_API_KEY'] !== undefined || env['OPENAI_BASE_URL'] !== undefined) {
      throw new HarnessError(
        'continue-cli OpenAI-compatible endpoints need a caller-selected configFile; the harness no longer generates Continue configs from OPENAI_API_KEY/OPENAI_BASE_URL',
        'unsupported-capability',
      )
    }

    let modelArgs: string[] = []
    if (model) {
      if (!isHubSlug(model)) {
        throw new HarnessError(
          `continue-cli --model takes a Continue Hub slug (owner/package), not the native model id ${JSON.stringify(model)}; pass a Hub slug, or leave model unset and select the model in a Continue config (configFile -> --config)`,
          'unsupported-capability',
        )
      }
      modelArgs = ['--model', model]
    }

    return finalizeCommand(this, spec, validated, {
      cmd: 'cn',
      args: ['-p', spec.prompt, ...permissionArgs, ...modelArgs, ...ruleArgs, '--format', 'json'],
      // Explicit Hub slug keeps its spelling; otherwise the upstream config picks.
      model: model ? spec.model! : null,
    })
  },

  parseOutput(_spec: RunSpec, outcome: SubprocOutcome): ParsedOutput {
    // Headless JSON is model-generated text; any usage/cost-looking fields in
    // it are not CLI telemetry.
    return { costUsd: null, tokensIn: null, tokensOut: null, raw: parseHeadlessJson(outcome.stdout) }
  },

  sessionLogPath(workdir: string, since?: number): string | null {
    return continueSessionPath(workdir, since)
  },

  parseSessionLog(path: string): SessionTelemetry {
    const usage = asRecord(readJsonObject(path)?.['usage'])
    if (usage === null) {
      return { sessionLogPath: path, tokensIn: null, tokensOut: null, costUsd: null, model: null, raw: null }
    }
    const tokensIn = numberOrNull(usage['promptTokens'])
    const tokensOut = numberOrNull(usage['completionTokens'])
    return {
      sessionLogPath: path,
      tokensIn: tokensIn === null ? null : Math.trunc(tokensIn),
      tokensOut: tokensOut === null ? null : Math.trunc(tokensOut),
      costUsd: numberOrNull(usage['totalCost']),
      model: null,
      raw: usage,
    }
  },
}

continueCliAdapter.submitKeys = ['Enter']
continueCliAdapter.detectReady = function (pane: string): ReadyState {
  const last20 = lastNonEmptyJoin(pane, 20)
  // cn shows "Ask anything" placeholder while model is still loading.
  // Real ready = input visible AND model loaded (no "Model: Loading..." line).
  if (/Model:\s*Loading/i.test(last20)) return 'loading'
  if (/Ask anything/i.test(last20)) return 'ready'
  if (/Update available/i.test(last20)) return 'dialog'
  return 'loading'
}
continueCliAdapter.handleDialog = function (pane: string): string[] | null {
  if (/Update available/i.test(stripAnsi(pane))) return ['Escape']
  return null
}
continueCliAdapter.detectStatus = function (pane: string): AgentStatus {
  const last10 = lastNonEmptyJoin(pane, 10)
  if (/rate.?limit/i.test(last10)) return 'rate-limited'
  if (/[⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏]/.test(last10)) return 'running'
  if (/thinking|working/i.test(last10)) return 'running'
  if (/Ask anything/i.test(last10)) return 'idle'
  return 'unknown'
}
continueCliAdapter.installMeta = {
  packageManager: 'npm',
  installCommand: ['npm', 'install', '-g', '@continuedev/cli'],
  updateCommand: ['npm', 'install', '-g', '@continuedev/cli@latest'],
  versionCommand: ['cn', '--version'],
}

register('continue-cli', continueCliAdapter)
